
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml" lang="en">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Commands from section ‘annotation’ &#8212; gtftk 1.0.1 documentation</title>
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Commands from section ‘coordinates’" href="coordinates.html" />
    <link rel="prev" title="Commands from section ‘conversion’" href="conversion.html" /> 
  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="coordinates.html" title="Commands from section ‘coordinates’"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="conversion.html" title="Commands from section ‘conversion’"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">gtftk 1.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="commands-from-section-annotation">
<h1>Commands from section ‘annotation’<a class="headerlink" href="#commands-from-section-annotation" title="Permalink to this headline">¶</a></h1>
<p>In the example of this section we will need the following example files:</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_example -q -d simple -f &#39;*&#39;
</pre></div>
</div>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk get_example -q -d mini_real -f &#39;*&#39;
</pre></div>
</div>
<hr class="docutils" />
<div class="section" id="ologram">
<h2>ologram<a class="headerlink" href="#ologram" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> OLOGRAM – OverLap Of Genomic Regions Analysis using Monte Carlo. Ologram annotates peaks
(in bed format) using (i) genomic features extracted from a GTF file (e.g promoter, tts, gene body, UTR…)
(ii) genomic regions tagged with particular keys/values in a GTF file (e.g. gene_biotype “protein_coding”,
gene_biotype “LncRNA”…) or (iii) from a BED file (e.g. user-defined regions). Each couple peak file/region
is randomly shuffled across the genome (inter-region lengths are considered). Then the probability of intersection
under the null hypothesis (the peaks and this feature are independent) is deduced thanks to this Monte Carlo approach.
The program will return statistics for both the number of intersections and the total lengths (in basepairs) of all intersections.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">The ologram examples below use 8 CPUs. Please adapt.</p>
</div>
<p><strong>Example:</strong> Perform a basic annotation. We are searching whether H3K4me3 peaks tends to be enriched in some specific genomic elements.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk ologram -i mini_real.gtf.gz -p ENCFF112BHN_H3K4me3_K562_sub.bed -c hg38.genome -u 1500 -d 1500 -D  -if example_pa_01.pdf -k 8
</pre></div>
</div>
<p>Download <a class="reference download internal" download="" href="_downloads/b84afafd9fd8ce1317545d9e1eba3df6/example_pa_01.pdf"><code class="xref download docutils literal notranslate"><span class="pre">pdf</span></code></a>.</p>
<br>
<table>
<tr>
<td valign="top">
<iframe src="_static/example_pa_01.pdf" title="your_title" align="top" width="500" height="620" width="50%" frameborder="0" scrolling="auto" target="Message">
</iframe>
</td>
</tr>
</table>
<br>
<br><p><strong>Example:</strong> Now we are using the gene_biotype key (note that a list of keys can be provided). This will tell us whether H3K4me3 tends to be located in particular transcripts (protein coding, LncRNAs…). The –no-basic-feature argument tells ologram not to test basic genomic elements (gene, transcripts…).</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk ologram -i mini_real.gtf.gz -m gene_biotype -p ENCFF112BHN_H3K4me3_K562_sub.bed -c hg38.genome -D -n  -if example_pa_02.pdf -k 8
 |-- 13:56-WARNING-ologram : Less than 2 remaining regions in one of the BED files. This is likely due to either : one of the considered features has very few peaks falling inside of it, or all the regions are in areas marked in the exclusion file. ologram will discard this particular pair.
 |-- 13:56-WARNING-ologram : Less than 2 remaining regions in one of the BED files. This is likely due to either : one of the considered features has very few peaks falling inside of it, or all the regions are in areas marked in the exclusion file. ologram will discard this particular pair.
 |-- 13:56-WARNING-ologram : Less than 2 remaining regions in one of the BED files. This is likely due to either : one of the considered features has very few peaks falling inside of it, or all the regions are in areas marked in the exclusion file. ologram will discard this particular pair.
 |-- 13:56-WARNING-ologram : Less than 2 remaining regions in one of the BED files. This is likely due to either : one of the considered features has very few peaks falling inside of it, or all the regions are in areas marked in the exclusion file. ologram will discard this particular pair.
 |-- 13:56-WARNING-ologram : Less than 2 remaining regions in one of the BED files. This is likely due to either : one of the considered features has very few peaks falling inside of it, or all the regions are in areas marked in the exclusion file. ologram will discard this particular pair.
</pre></div>
</div>
<p>Download <a class="reference download internal" download="" href="_downloads/6a22a2c6705c1421955406a0316272f5/example_pa_02.pdf"><code class="xref download docutils literal notranslate"><span class="pre">pdf</span></code></a>.</p>
<br>
<table>
<tr>
<td valign="top">
<iframe src="_static/example_pa_02.pdf" title="your_title" align="top" width="500" height="620" width="50%" frameborder="0" scrolling="auto" target="Message">
</iframe>
</td>
</tr>
</table>
<br>
<br><p><strong>Example:</strong> A more complex example where the key is created on the fly. Expression data are loaded as a novel key using the join_attr command and associated to gene features. This novel key (exprs) is then discretized to created 6 classes of genes with increasing expression (based on percentiles, -p) which are tested for enrichment in H3K4me3.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk join_attr -i mini_real.gtf.gz -H -j mini_real_counts_ENCFF630HEX.tsv -k gene_name -n exprs -t gene | gtftk discretize_key -k exprs -p -d exprs_class -n 6   | gtftk ologram -p ENCFF112BHN_H3K4me3_K562_sub.bed -c hg38.genome -D -n -m exprs_class -if example_pa_03.pdf -k 8
 |-- 13:57-INFO-discretize_key : Categories: [&#39;[0.0_10.0]&#39;, &#39;(10.0_149.0]&#39;, &#39;(149.0_585.0]&#39;, &#39;(585.0_1171.0]&#39;, &#39;(1171.0_2486.83]&#39;, &#39;(2486.83_41703.0]&#39;]
</pre></div>
</div>
<p>Download <a class="reference download internal" download="" href="_downloads/959e411f4470e8e1330736497073c14d/example_pa_03.pdf"><code class="xref download docutils literal notranslate"><span class="pre">pdf</span></code></a>.</p>
<br>
<table>
<tr>
<td valign="top">
<iframe src="_static/example_pa_03.pdf" title="your_title" align="top" width="500" height="620" width="50%" frameborder="0" scrolling="auto" target="Message">
</iframe>
</td>
</tr>
</table>
<br>
<br><p><strong>Example:</strong> Using the add_exon_nb, we add the exon number transcript-wise (numbering from 5’ to 3’) and discretize this novel key into 5 classes tested for enrichment.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk add_exon_nb -k exon_nbr -i mini_real.gtf.gz | gtftk discretize_key -p -d exon_nbr_cat -n 5  -k exon_nbr | gtftk ologram -p ENCFF112BHN_H3K4me3_K562_sub.bed -c hg38.genome -D -n -m exon_nbr_cat -if example_pa_04.pdf -k 8
 |-- 13:58-INFO-discretize_key : Categories: [&#39;[1.0_2.0]&#39;, &#39;(2.0_4.0]&#39;, &#39;(4.0_6.0]&#39;, &#39;(6.0_12.0]&#39;, &#39;(12.0_107.0]&#39;]
</pre></div>
</div>
<p>Download <a class="reference download internal" download="" href="_downloads/614b25d8398e8c669baa9f24c80a4a3f/example_pa_04.pdf"><code class="xref download docutils literal notranslate"><span class="pre">pdf</span></code></a>.</p>
<br>
<table>
<tr>
<td valign="top">
<iframe src="_static/example_pa_04.pdf" title="your_title" align="top" width="500" height="620" width="50%" frameborder="0" scrolling="auto" target="Message">
</iframe>
</td>
</tr>
</table>
<br>
<br><p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk ologram -h
  Usage: gtftk ologram [-i GTF] [-c TXT] -p BED [-b [more_bed [more_bed ...]]] [-l more_bed_labels] [-e BED] [-u upstream] [-d downstream] [-m more_keys] [-n] [-k nb_threads] [-s seed] [-mn minibatch_nb] [-ms minibatch_size] [-ma] [-o DIR] [-pw pdf_width] [-ph pdf_height] [-if user_img_file] [-x] [-tp tsv_file_path] [-dpi dpi] [-z] [-f] [-w] [-q] [-h] [-V ] [-D] [-C] [-K] [-A] [-L]

  Description: 
     *  OLOGRAM -- OverLap Of Genomic Regions Analysis using Monte Carlo. Ologram  annotates peaks (in bed
     format) using (i) genomic features extracted  from a GTF file (e.g promoter, tts, gene body,
     UTR...) (ii) genomic regions tagged with  particular keys/values in a GTF file (e.g.
     gene_biotype &quot;protein_coding&quot;,  gene_biotype &quot;LncRNA&quot;...) or (iii) from a BED file (e.g. user-
     defined regions).  Each couple peak file/region is randomly shuffled across the genome (inter-
     region lengths are considered). Then the probability of intersection under the null hypothesis
     (the peaks and this feature are independent) is deduced thanks to this Monte Carlo approach.
     The program will return statistics for both the number of intersections and the total lengths
     (in basepairs) of all intersections.  Authors : Quentin Ferré &lt;quentin.q.ferre@gmail.com&gt; and
     Denis Puthier &lt;denis.puthier@univ-amu.fr&gt;.

  Notes:
     *  Although ologram itself is not RAM-intensive, base pygtftk processing of a full human GTF
     can require upwards of 8Gb. It is recommended you do not run other programs in the meantime on
     a laptop.
     *  Genome size is computed from the provided chromInfo file (-c). It should thus only contain
     ordinary chromosomes.
     *  --chrom-info may also accept &#39;mm8&#39;, &#39;mm9&#39;, &#39;mm10&#39;, &#39;hg19&#39;, &#39;hg38&#39;, &#39;rn3&#39; or &#39;rn4&#39;. In this
     case the corresponding size of conventional chromosomes are used. ChrM is not used.
     *  The program produces a pdf file and a txt file (&#39;_stats_&#39;) containing intersection
     statistics for the shuffled BEDs under H0 (peak_file and the considered genomic region are
     independant): number of intersections (= number of lines in the bed intersect) and total number
     of overlapping base pairs. The output figure gives, for both statistics, esperance and standard
     deviation (error bars) in the shuffles compared to the actual values. It also gives, under the
     &#39;fit&#39; label for each statistic, the goodness of fit of the statistic under (H0) to a Negative
     Binomial assessed by a Cramer&#39;s V score (fit_quality gives 1-V ; as per Cramer (1948) a good
     fit should have a fit quality above (1 - 0.25 = 0.75) if your nb. of shuffles is in the
     hundreds, but closer to 0.9 if it is in the thousands or above.  The p-value of the true
     intersection under the distribution characterized by the shuffles is also given, under
     &#39;p_value&#39;. Finally, the log2 fold change between true and shuffles is also given.
     *  If --more-keys is used additional region sets will be tested based on the associated key
     value. As an example, if --more-keys is set to the &#39;gene_biotype&#39; (a key generally found in
     ensembl GTF), the region related to &#39;protein_coding&#39;, &#39;lncRNA&#39; or any other values for that key
     will be retrieved merged and tested for enrichment.
     *  Use --no-basic-feature if you want to perform enrichment analysis on focused annotations
     only (--more-bed or --more-key).
     *  The goal of the minibatch is to save RAM. Increase the number of minibatches, instead of
     their size. You may need to use very small minibatches if you have large sets of regions.
     *  You can exclude regions from the shuffling. This is done by shuffling across a concatenated
     &quot;sub-genome&quot; obtained by removing the excluded regions, but the same ones will be excluded from
     the peak_file and the GTF. This in Beta for now and will be very time-consuming (hours),
     especially if you have few CPU cores. Try using an exclusion file that is as small (around a
     thousand elements) as possible.
     *  BETA : About --use-markov. This arguments control whether to use Markov model realisations
     instead of independant shuffles for respectively region lengths and inter-region lengths. This
     is not recommended in the general case and can *very* time-consuming (hours).

  Version:  2019-03-18

Arguments:
 -i, --inputfile             Path to the GTF file. Defaults to STDIN (default: &lt;stdin&gt;)
 -c, --chrom-info            Tabulated two-columns file. Chromosomes as column 1, sizes as column 2 (default: None)
 -p, --peak-file             The file containing the peaks/regions to be annotated. (bed format). (default: None)
 -b, --more-bed              A list of bed files to be considered as additional genomic annotations. (default: None)
 -l, --more-bed-labels       A comma separated list of labels (see --more-bed) (default: None)
 -e, --bed-excl              Exclusion file. The chromosomes will be shortened by this much for the shuffles of peaks and features. Can take a long time. (bed format). (default: None)
 -u, --upstream              Extend the TSS and TTS of in 5&#39; by a given value. (default: 1000)
 -d, --downstream            Extend the TSS and TTS of in 3&#39; by a given value. (default: 1000)
 -m, --more-keys             A comma separated list of key used for labeling the genome. See Notes. (default: None)
 -n, --no-basic-feature      No statistics for basic features of GTF. Concentrates on --more-bed and --more-keys. (default: False)
 -k, --nb-threads            Number of threads for multiprocessing. (default: 8)
 -s, --seed                  Numpy random seed. (default: 42)
 -mn, --minibatch-nb         Number of minibatches of shuffles. (default: 10)
 -ms, --minibatch-size       Size of each minibatch, in number of shuffles. (default: 20)
 -ma, --use-markov           Whether to use Markov model realisations instead of independant shuffles. See notes. (default: False)
 -o, --outputdir             Output directory name. (default: ologram_output)
 -pw, --pdf-width            Output pdf file width (inches). (default: None)
 -ph, --pdf-height           Output pdf file height (inches). (default: None)
 -if, --user-img-file        Provide an alternative path for the main image. (default: None)
 -x, --no-pdf                Do not produce any image file. (default: False)
 -tp, --tsv-file-path        Provide an alternative path for text output file. (default: None)
 -dpi, --dpi                 Dpi to use. (default: 300)
 -z, --no-gtf                No gtf file is provide as input. (default: False)
 -f, --force-chrom-gtf       Discard silently, from GTF, genes outside chromosomes defined in --chrom-info. (default: False)
 -w, --force-chrom-peak      Discard silently, from --peak-file, peaks outside chromosomes defined in --chrom-info. (default: False)
 -q, --force-chrom-more-bed  Discard silently, from --more-bed files, regions outside chromosomes defined in --chrom-info. (default: False)

Command-wise optional arguments:
 -h, --help                  Show this help message and exit.
 -V, --verbosity             Increase output verbosity. (default: 0)
 -D, --no-date               Do not add date to output file names. (default: False)
 -C, --add-chr               Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir               Keep all temporary files into this folder. (default: None)
 -A, --keep-all              Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file           Stores the arguments passed to the command into a file. (default: None)
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="closest-genes">
<h2>closest_genes<a class="headerlink" href="#closest-genes" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Find the n closest genes for each transcript.</p>
<p><strong>Example:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk closest_genes  -i simple.gtf -f
genes	closest_genes	distances
G0001	G0007	18
G0002	G0010	4
G0003	G0004	4
G0004	G0003	4
G0005	G0006	12
G0006	G0005	12
G0007	G0001	18
G0008	G0002	42
G0009	G0006	21
G0010	G0002	4
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk closest_genes -h
  Usage: gtftk closest_genes [-i GTF] [-o GTF/TXT] [-r {tss,tts,gene}] [-nb nb_neighbors] [-t {tss,tts,gene}] [-s] [-S] [-f] [-H] [-k] [-id {gene_id,gene_name}] [-h] [-V ] [-D] [-C] [-K] [-A] [-L]

  Description: 
     *  Find the n closest genes for each genes.

  Notes:
     *  The reference region for each gene can be the TSS (the most 5&#39;), the TTS (The most 3&#39;) or
     the whole gene.
     *  The reference region for each closest gene can be the TSS, the whole gene or the TTS.
     *  The closest genes can be searched in a stranded or unstranded fashion.

  Version:  2018-02-11

optional arguments:
 -i, --inputfile          Path to the GTF file. Default to STDIN (default: &lt;stdin&gt;)
 -o, --outputfile         Output file. (default: &lt;stdout&gt;)

Arguments:
 -r, --from-region-type   What is region to consider for each gene. (default: tss)
 -nb, --nb-neighbors      The size of the neighborhood. (default: 1)
 -t, --to-region-type     What is region to consider for each closest gene. (default: tss)
 -s, --same-strandedness  Require same strandedness (default: False)
 -S, --diff-strandedness  Require different strandedness (default: False)
 -f, --text-format        Return a text format. (default: False)
 -H, --no-header          Don&#39;t print the header line. (default: False)
 -k, --collapse           Unwrap. Don&#39;t use comma. Print closest genes line by line. (default: False)
 -id, --identifier        The key used as gene identifier. (default: gene_id)

Command-wise optional arguments:
 -h, --help               Show this help message and exit.
 -V, --verbosity          Increase output verbosity. (default: 0)
 -D, --no-date            Do not add date to output file names. (default: False)
 -C, --add-chr            Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir            Keep all temporary files into this folder. (default: None)
 -A, --keep-all           Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file        Stores the arguments passed to the command into a file. (default: None)
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="overlapping">
<h2>overlapping<a class="headerlink" href="#overlapping" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Find transcripts whose body/TSS/TTS region extended in 5’ and 3’ (-u/-d) overlaps with any transcript from another gene. Strandness is not considered by default. Used –invert-match to find those that do not overlap. If –annotate-gtf is used, all lines of the input GTF file will be printed and a new key containing the list of overlapping transcripts will be added to the transcript features/lines (key will be ‘overlapping_*’ with * one of body/TSS/TTS). The –annotate-gtf and –invert-match arguments are mutually exclusive.</p>
<p><strong>Example:</strong> Find transcript whose promoter overlap transcript from other genes.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk overlapping -i simple.gtf -c simple.chromInfo -t promoter -u 10 -d 10 -a    | gtftk select_by_key -k feature -v transcript | gtftk tabulate -k transcript_id,overlap_promoter_u0.01k_d0.01k | head
transcript_id	overlap_promoter_u0.01k_d0.01k
G0001T002	G0007T001,G0007T002
G0001T001	G0007T001,G0007T002
G0002T001	G0010T001
G0003T001	G0004T002,G0004T001
G0004T002	G0003T001
G0004T001	G0003T001
G0005T001	G0003T001
G0006T001	G0005T001
G0006T002	G0005T001
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk overlapping -h
  Usage: gtftk overlapping [-i GTF] [-o GTF] -c CHROMINFO [-u UPSTREAM] [-d DOWNSTREAM] [-t {transcript,promoter,tts}] [-s] [-S] [-n] [-a] [-k key_name] [-h] [-V ] [-D] [-C] [-K] [-A] [-L]

  Description: 
     *  Find transcripts whose body/TSS/TTS region extended in 5&#39; and 3&#39; (-u/-d) overlaps with any
     transcript from another gene. Strandness is not considered by default. Used --invert-match to
     find those that do not overlap. If --annotate-gtf is used, all lines of the input GTF file will
     be printed and a new key containing the list of overlapping transcripts will be added to the
     transcript features/lines (key will be &#39;overlapping_*&#39; with * one of body/TSS/TTS). The
     --annotate-gtf and --invert-match arguments are mutually exclusive.

  Notes:
     *  --chrom-info may also accept &#39;mm8&#39;, &#39;mm9&#39;, &#39;mm10&#39;, &#39;hg19&#39;, &#39;hg38&#39;, &#39;rn3&#39; or &#39;rn4&#39;. In this
     case the  corresponding size of conventional chromosomes are used. ChrM is not used.

  Version:  2018-01-24

Arguments:
 -i, --inputfile          Path to the GTF file. Default to STDIN (default: &lt;stdin&gt;)
 -o, --outputfile         Output file. (default: &lt;stdout&gt;)
 -c, --chrom-info         Chromosome information. A tabulated two-columns file with chromosomes as column 1 and sizes as column 2 (default: None)
 -u, --upstream           Extend the region in 5&#39; by a given value (int). Used to define the region around the TSS/TTS. (default: 1500)
 -d, --downstream         Extend the region in 3&#39; by a given value (int). Used to define the region around the TSS/TTS. (default: 1500)
 -t, --feature-type       The feature of interest. (default: transcript)
 -s, --same-strandedness  Require same strandedness (default: False)
 -S, --diff-strandedness  Require different strandedness (default: False)
 -n, --invert-match       Not/Invert match. (default: False)
 -a, --annotate-gtf       All lines of the original GTF will be printed. (default: False)
 -k, --key-name           The name of the key. (default: None)

Command-wise optional arguments:
 -h, --help               Show this help message and exit.
 -V, --verbosity          Increase output verbosity. (default: 0)
 -D, --no-date            Do not add date to output file names. (default: False)
 -C, --add-chr            Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir            Keep all temporary files into this folder. (default: None)
 -A, --keep-all           Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file        Stores the arguments passed to the command into a file. (default: None)
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="divergent">
<h2>divergent<a class="headerlink" href="#divergent" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Find transcript with divergent promoters. These transcripts will be defined here
as those whose promoter region (defined by -u/-d) overlaps with the tss of
another gene in reverse/antisens orientation. This may be useful to select
coding genes in head-to-head orientation or LUAT as described in “Divergent
transcription is associated with promoters of transcriptional regulators”
(Lepoivre C, BMC Genomics, 2013). The output is a GTF with an additional key
(‘divergent’) whose value is set to ‘.’ if the gene has no antisens transcript
in its promoter region. If the gene has an antisens transcript in its promoter
region the ‘divergent’ key is set to the identifier of the transcript whose tss
is the closest relative to the considered promoter. The tss to tss distance is
also provided as an additional key (dist_to_divergent).</p>
<p><strong>Example:</strong> Flag divergent transcripts in the example dataset. Select them and produce a tabulated output.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk divergent -i simple.gtf -c simple.chromInfo -u 10 -d 10| gtftk select_by_key -k feature -v transcript | gtftk tabulate -k transcript_id,divergent,dist_to_divergent | head  -n 7
transcript_id	divergent	dist_to_divergent
G0003T001	G0004T002	4
G0004T002	G0003T001	4
G0004T001	G0003T001	4
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk divergent -h
  Usage: gtftk divergent [-i GTF] [-o GTF] -c CHROMINFO [-u UPSTREAM] [-d DOWNSTREAM] [-n] [-S] [-a key_name] [-h] [-V ] [-D] [-C] [-K] [-A] [-L]

  Description: 
     *  Find transcripts with divergent promoters. These transcripts will be defined here as those whose
     promoter region (defined by -u/-d) overlaps with the tss of another gene in reverse/antisens
     orientation. This may be useful to select coding genes in head-to-head orientation or LUAT as
     described in &quot;Divergent transcription is associated with promoters of transcriptional
     regulators&quot; (Lepoivre C, BMC Genomics, 2013). The output is a GTF with an additional key
     (&#39;divergent&#39;) whose value is set to &#39;.&#39; if the gene has no antisens transcript in its promoter
     region. If the gene has an antisens transcript in its promoter region the &#39;divergent&#39; key is
     set to the identifier of the transcript whose tss is the closest relative to the considered
     promoter. The tss to tss distance is also provided as an additional key (dist_to_divergent).

  Notes:
     *  --chrom-info may also accept &#39;mm8&#39;, &#39;mm9&#39;, &#39;mm10&#39;, &#39;hg19&#39;, &#39;hg38&#39;, &#39;rn3&#39; or &#39;rn4&#39;. In this
     case the  corresponding size of conventional chromosomes are used. ChrM is not used.

  Version:  2018-01-24

Arguments:
 -i, --inputfile      Path to the GTF file. Default to STDIN (default: &lt;stdin&gt;)
 -o, --outputfile     Output file. (default: &lt;stdout&gt;)
 -c, --chrom-info     Tabulated two-columns file. Chromosomes as column 1 and their sizes as column 2 (default: None)
 -u, --upstream       Extend the promoter in 5&#39; by a given value (int). Defines the region around the tss. (default: 1500)
 -d, --downstream     Extend the region in 3&#39; by a given value (int). Defines the region around the tss. (default: 1500)
 -n, --no-annotation  Do not annotate the GTF. Just select the divergent transcripts. (default: False)
 -S, --no-strandness  Do not consider strandness (only look whether the promoter from a transcript overlaps with the promoter from another gene). (default: False)
 -a, --key-name       The name of the key. (default: None)

Command-wise optional arguments:
 -h, --help           Show this help message and exit.
 -V, --verbosity      Increase output verbosity. (default: 0)
 -D, --no-date        Do not add date to output file names. (default: False)
 -C, --add-chr        Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir        Keep all temporary files into this folder. (default: None)
 -A, --keep-all       Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file    Stores the arguments passed to the command into a file. (default: None)
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="convergent">
<h2>convergent<a class="headerlink" href="#convergent" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Find transcript with convergent tts. These transcripts will be defined here
as those whose tts region (defined by -u/-d) overlaps with the tts of
another gene in reverse/antisens orientation. The output is a GTF with an
additional key (‘convergent’) whose value is set to ‘.’ if the gene has no
convergent transcript in its tts region. If the gene has an antisens transcript
in its tts region the ‘convergent’ key is set to the identifier of the
transcript whose tts is the closest relative to the considered tts.
The tts to tts distance is also provided as an additional key (dist_to_convergent).</p>
<p><strong>Example:</strong> Flag divergent transcripts in the example dataset. Select them and produce a tabulated output.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk convergent -i simple.gtf -c simple.chromInfo -u 25 -d 25| gtftk select_by_key -k feature -v transcript | gtftk tabulate -k transcript_id,convergent,dist_to_convergent| head -n 4
transcript_id	convergent	dist_to_convergent
G0002T001	G0008T001	21
G0008T001	G0002T001	21
G0010T001	G0008T001	24
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk convergent -h
  Usage: gtftk convergent [-i GTF] [-o GTF] -c CHROMINFO [-u UPSTREAM] [-d DOWNSTREAM] [-h] [-V ] [-D] [-C] [-K] [-A] [-L]

  Description: 
     *  Find transcripts with convergent tts. These transcripts will be defined here as those whose tts
     region (defined by -u/-d) overlaps with the tts of another gene in reverse/antisens
     orientation. The output is a GTF with an additional key (&#39;convergent&#39;) whose value is set to
     &#39;.&#39; if the gene has no convergent transcript in its tts region. If the gene has an antisens
     transcript in its tts region the &#39;convergent&#39; key is set to the identifier of the transcript
     whose tts is the closest relative to the considered tts. The tts to tts distance is also
     provided as an additional key (dist_to_convergent).

  Notes:
     *  --chrom-info may also accept &#39;mm8&#39;, &#39;mm9&#39;, &#39;mm10&#39;, &#39;hg19&#39;, &#39;hg38&#39;, &#39;rn3&#39; or &#39;rn4&#39;. In this
     case the  corresponding size of conventional chromosomes are used. ChrM is not used.

  Version:  2018-01-20

Arguments:
 -i, --inputfile    Path to the GTF file. Default to STDIN (default: &lt;stdin&gt;)
 -o, --outputfile   Output file. (default: &lt;stdout&gt;)
 -c, --chrom-info   Tabulated two-columns file. Chromosomes as column 1 and sizes as column 2 (default: None)
 -u, --upstream     Extends the tts in 5&#39; by a given value (int). Defines the region around the tts. (default: 1500)
 -d, --downstream   Extends the region in 3&#39; by a given value (int). Defines the region around the tts. (default: 1500)

Command-wise optional arguments:
 -h, --help         Show this help message and exit.
 -V, --verbosity    Increase output verbosity. (default: 0)
 -D, --no-date      Do not add date to output file names. (default: False)
 -C, --add-chr      Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir      Keep all temporary files into this folder. (default: None)
 -A, --keep-all     Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file  Stores the arguments passed to the command into a file. (default: None)
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="exon-sizes">
<h2>exon_sizes<a class="headerlink" href="#exon-sizes" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Add a new key to transcript features containing a comma-separated list of exon sizes.</p>
<p><strong>Example:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk exon_sizes -i simple.gtf | gtftk select_by_key -t | gtftk tabulate -k transcript_id,exon_sizes
transcript_id	exon_sizes
G0001T002	14
G0001T001	14
G0002T001	10
G0003T001	5,5
G0004T002	4,1,3
G0004T001	4,1,3
G0005T001	6,3
G0006T001	3,3,4
G0006T002	3,3
G0007T001	10
G0007T002	10
G0008T001	3,5
G0009T002	12
G0009T001	12
G0010T001	11
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk exon_sizes -h
  Usage: gtftk exon_sizes [-i GTF] [-o TXT] [-a key_name] [-h] [-V ] [-D] [-C] [-K] [-A] [-L]

  Description: 
     *  Add a new key to transcript features containing a comma-separated list of exon sizes.

  Version:  2018-01-24

Arguments:
 -i, --inputfile    Path to the GTF file. Default to STDIN. (default: &lt;stdin&gt;)
 -o, --outputfile   Output GTF file. (default: &lt;stdout&gt;)
 -a, --key-name     The name of the key. (default: exon_sizes)

Command-wise optional arguments:
 -h, --help         Show this help message and exit.
 -V, --verbosity    Increase output verbosity. (default: 0)
 -D, --no-date      Do not add date to output file names. (default: False)
 -C, --add-chr      Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir      Keep all temporary files into this folder. (default: None)
 -A, --keep-all     Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file  Stores the arguments passed to the command into a file. (default: None)
</pre></div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="intron-sizes">
<h2>intron_sizes<a class="headerlink" href="#intron-sizes" title="Permalink to this headline">¶</a></h2>
<p><strong>Description:</strong> Add a new key to transcript features containing a comma-separated list of intron sizes.</p>
<p><strong>Example:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk intron_sizes -i simple.gtf | gtftk select_by_key -t | gtftk tabulate -k transcript_id,intron_sizes
transcript_id	intron_sizes
G0001T002	0
G0001T001	0
G0002T001	0
G0003T001	2
G0004T002	2,2
G0004T001	2,2
G0005T001	6
G0006T001	2,2
G0006T002	2
G0007T001	0
G0007T002	0
G0008T001	5
G0009T002	0
G0009T001	0
G0010T001	0
</pre></div>
</div>
<p><strong>Arguments:</strong></p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>$ gtftk intron_sizes -h
  Usage: gtftk intron_sizes [-i GTF] [-o GTF] [-a key_name] [-h] [-V ] [-D] [-C] [-K] [-A] [-L]

  Description: 
     *  Add a new key to transcript features containing a comma-separated list of intron-size.

  Version:  2018-01-24

Arguments:
 -i, --inputfile    Path to the GTF file. Default to STDIN. (default: &lt;stdin&gt;)
 -o, --outputfile   Output file. (default: &lt;stdout&gt;)
 -a, --key-name     The name of the key. (default: intron_sizes)

Command-wise optional arguments:
 -h, --help         Show this help message and exit.
 -V, --verbosity    Increase output verbosity. (default: 0)
 -D, --no-date      Do not add date to output file names. (default: False)
 -C, --add-chr      Add &#39;chr&#39; to chromosome names before printing output. (default: False)
 -K, --tmp-dir      Keep all temporary files into this folder. (default: None)
 -A, --keep-all     Try to keep all temporary files even if process does not terminate normally. (default: False)
 -L, --logger-file  Stores the arguments passed to the command into a file. (default: None)
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Commands from section ‘annotation’</a><ul>
<li><a class="reference internal" href="#ologram">ologram</a></li>
<li><a class="reference internal" href="#closest-genes">closest_genes</a></li>
<li><a class="reference internal" href="#overlapping">overlapping</a></li>
<li><a class="reference internal" href="#divergent">divergent</a></li>
<li><a class="reference internal" href="#convergent">convergent</a></li>
<li><a class="reference internal" href="#exon-sizes">exon_sizes</a></li>
<li><a class="reference internal" href="#intron-sizes">intron_sizes</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="conversion.html"
                        title="previous chapter">Commands from section ‘conversion’</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="coordinates.html"
                        title="next chapter">Commands from section ‘coordinates’</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/annotation.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="coordinates.html" title="Commands from section ‘coordinates’"
             >next</a> |</li>
        <li class="right" >
          <a href="conversion.html" title="Commands from section ‘conversion’"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">gtftk 1.0.1 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2018, F. Lopez and D. Puthier.
      Last updated on Mar 22, 2019.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.8.3.
    </div>
  </body>
</html>